extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -45,34 +45,128 @@
+            if (i == currentFrame.upperBoundExclusive()) {
+                cachedNonNullIndex = null;
+                break;
             }
-        },
-        BACKWARD {
-            @Override
-            int getTargetIndex(int idxInPartition, int offset) {
-                return idxInPartition - offset;
+            Object value = getValueAtTargetIndex(i, currentFrame, expressions, args);
+            if (value != null) {
+                cachedNonNullIndex = i;
+                break;
+            }
+        }
+    }
+
+    private Integer findNonNullOffsetFromCurrentIndex(int idxInPartition,
+                                                      int offset,
+                                                      WindowFrameState currentFrame,
+                                                      List<? extends CollectExpression<Row, ?>> expressions,
+                                                      Input[] args) {
+        /* Search for 'offset' number of non-null elements in 'resolvedDirection' and return the index if found.
+           If index goes out of bound, an exception will be thrown and eventually invoke getDefaultOrNull(...) */
+        for (int i = 1, counter = 0; ; i++) {
+            Object value = getValueAtTargetIndex(getTargetIndex(idxInPartition, i), currentFrame, expressions, args);
+            if (value != null) {
+                counter++;
+                if (counter == offset) {
+                    return getTargetIndex(idxInPartition, i);
+                }
             }
-        };
-        abstract int getTargetIndex(int idxInPartition, int offset);
+        }
+    }
+
+    private Object getValueAtOffset(int idxAtPartition,
+                                    int offset,
+                                    WindowFrameState currentFrame,
+                                    List<? extends CollectExpression<Row, ?>> expressions,
+                                    boolean ignoreNulls,
+                                    Input[] args) {
+        if (ignoreNulls == false) {
+            var targetIndex = getTargetIndex(idxAtPartition, offset);
+            return getValueAtTargetIndex(targetIndex, currentFrame, expressions, args);
+        } else {
+            return getValueAtOffsetIgnoringNulls(idxAtPartition, offset, currentFrame, expressions, args);
+        }
+    }
+
+    private static Object getValueAtTargetIndex(Integer targetIndex,
+                                                WindowFrameState currentFrame,
+                                                List<? extends CollectExpression<Row, ?>> expressions,
+                                                Input[] args) {
+        if (targetIndex == null) {
+            throw new IndexOutOfBoundsException();
+        }
+        Object[] rowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
+        if (rowCells == null) {
+            throw new IndexOutOfBoundsException();
+        }
+        var row = new RowN(rowCells);
```

Comment:
I think it might be worth creating a re-usable row instance on the class level. `getValueAtTargetIndex` is called within a loop so this here might cause unnecessary allocations. (depending on whether escape analysis would work or not)

Reply:
How about modifying `WindowFrameState.rows` from `List<Object[]>` to `List<Row>` ? It is seemed to be common that the window function implementations convert it to `Row` right away

Reply:
That would increase the footprint by ~24 bytes per row. Depending on the framing settings, I think there can be a large number of rows in it, so changing this could have quite an impact.